Replace default application layout and homepage (currently Browse Projects)

amclark42 opened the pull request:
This PR contains updates drawn from the TAPAS wireframes:

- CSS and Javascript
- a modified application layout for every page, including
    - header,
    - footer, and 
    - user login/menu
- a new Browse Projects page, which also serves as the homepage for now (replacing the Blacklight catalog); this consists of
    - a Project-specific browse template
    - a generic browse template
    - a Project-specific browse "container" (list item)
    - a generic partial for a button that collapses or expands some content (making it easier to reproduce its accessible features)

### Testing the PR

- **Web display**
    - Does the homepage load as expected?
    - Are there any places where the TAPAS layout isn't working, or the page content is negatively affected?
- **Code quality**
    - Have I made any errors?
    - Are there things I can improve? (syntax, Rails-standard ways of doing things, etc.)

### Background

I originally planned to submit a PR with just the web assets and application layout. However, the Blacklight catalog page is incompatible with the XHTML format of our new layout — I felt I should create a replacement before sharing my work.

While creating the Browse Projects page templates, I found that the Project methods `collections` and `core_files` (intended to return those resources connected to the current project) were likely making inefficient database queries — the Browse page took around a minute to load due to the `core_files` method alone. Both methods have been commented out in favor of the default ones we can use because of the models' defined relationships.

I also set up a `browse` method in the Projects controller. I used [the `:includes` method](https://engineering.gusto.com/a-visual-guide-to-using-includes-in-rails-700a91cd3095) to have the relevant collections and core files ready for use in the page templates.

Unfortunately, Rails doesn't recognize that `core_files` are directly attached to the Project at all (I think @candyhazlett said that they were attached manually on her branch?). Currently, the Browse page lists 0 core files for each project. But since the page loads snappily and we need time to work out the models' relationships, we felt this was a good time for me to submit a PR.

### 2025-08 edit: Additional context

This pull request contains work with two goals:

1. Recreate the TAPAS page structure from the TAPAS wireframes repository.
    1. Add CSS, JS, and icons from that repository.
2. Recreate "Browse Projects" wireframe as a list of all TAPAS projects.

As noted in my original description, upon making goal 1 happen, I discovered that switching to XHTML output caused the Blacklight catalog Javascript to stop working. I knew that Candace relied on the Browse page for testing, so I hurriedly implemented the "Browse Projects" page as a replacement for the Blacklight page. (To be clear, we were always going to move away from the Blacklight catalog JS, but its breakage accelerated the need for a replacement.)

The new interface requires only Bootstrap v5+, Bootstrap's dependency Popper, and TAPAS-specific Javascripts. JQuery (and many other JS libraries) are no longer necessary — so the interface no longer loads them.

#### Scope of the work

The "Browse Projects" page here is in [the minimal form shown in the wireframe](https://tapasproject.org/wireframes/browse/projects.html). It **does**:

- List every project in TAPAS Rails,
- with expandable containers to get at more information:
    - date last updated, 
    - description, 
    - project admins, 
    - number of collections, and 
    - number of TEI records.

The [specification for "Browse Projects"](https://docs.google.com/document/d/1vsD_ZJtoHAQo5-T-frH3w64Tpbmuyp4y_JaVu01Hc6w/edit?tab=t.0#heading=h.v90ktn7h1hk4) pre-dates the page wireframe and discussions about it — it has not yet been updated to reflect group feedback. The spec currently states:
> **See a list or grid view of all TAPAS projects**
>   Sortable and filterable by owner, by title, by date created, by number of TEI files
>   Grid view shows thumbnail image (or default TAPAS thumbnail) and title

Emphasis is mine. The "Browse Projects" page is currently "just" a list of all TAPAS projects. The page does **not** include:

- Sorting. The sort controls are present on the page but don't yet do anything.
- Pagination.
- Filtering. The TAPAS strategy team decided in April not to implement filtering (since we have few fields which would be useful to filter on).
- Grid view. Not discussed in a strategy meeting; this may be a nice-to-have feature.

amclark42 committed:
Create folder for web assets; add view packages submodule

amclark42 committed:
CSS tweaks; update paths to web assets; include TAPAS Generic content

amclark42 committed:
Add TAPAS Generic CSS and JS to the reader wireframe.

The dialog boxes are partially-hidden behind the header. I'm not
sure why the absolute positioning works in Prod, but it definitely
doesn't work here. Might need to update the view package.

amclark42 committed:
Make the Reader header sticky, not the TAPAS header.

The dropdown to choose a view package now stands out a little more.

amclark42 committed:
Make the jump-to-top link sticky when the header is not in view.

I used the JS Intersection Observer for the first time!

amclark42 committed:
Adjust scroll button intersection observer: test the reader header.

Instead of testing the *page* header, we test the reader header for
an intersection threshold of 1.0. Thanks to a new `top: -1px` rule,
we now know that when the reader header is sticky, a part of it is
just out of the viewport.

I also set up an `.is-sticking` class on the reader header,
preparing to maybe reduce the amount of space the sticky header
takes up.

amclark42 committed:
Edit JS comments

amclark42 committed:
When the reader header sticks, truncate long document titles

amclark42 committed:
Use CSS Grid to lay out Reader header.

The header is now laid out for both desktop and mobile screens. It
does still take up a lot of space in sticky form, but I can work on
that now that the structure and base layout is solid.

amclark42 committed:
Don't truncate long titles when the Reader header sticks.

Truncating long titles led to jittering when scrolling down, or
jumping to the top of the page. Still need a way to make the
"stuck" content take up less of the screen.

amclark42 committed:
Only apply sticky positioning on "medium" screens or larger.

When screen space is at a premium, there's no need to take up more
by introducing a sticky header.

Also, make sure that the contents of the Reader don't overflow
their container. Horizontal scroll isn't great, but the TAPAS
interface needs to be able to handle problematic view packages.

amclark42 committed:
Keep "Return to top" from showing when the header is visible.

With a long title and description, the reader header sometimes
clipped off the viewport on very small screens, which meant that
the intersection ratio was < 1 and the "Return to top" button would
show up. Now, both breadcrumbs and reader header are observed for
visibility, and "Return to top" is only shown when it is possible
to scroll up to the `#main-content`.

amclark42 committed:
Use <details> for reader doc description.

Adjusted styling of `.reader-desc` to better communicate what's
going on, though I'm not certain about the color.

Ideally we can toggle the widget closed when the reader header is
in sticky mode.

amclark42 committed:
Adjust border radius depending on screen width

amclark42 committed:
Toggle doc description closed when reader header is sticky.

To avoid visual glitches, <details> isn't toggled open
automatically; the user will have to toggle it manually. I think
this is acceptable.

amclark42 committed:
Prevent the description from being closed when the page loads.

If the page was reloaded when the reader header is in sticky
position, we don't want the description to flicker open and then
closed. Rather, the description should stay open until the user
either closes it themself, or scrolls up and back down.

amclark42 committed:
Handle case where description doesn't exist

amclark42 committed:
Small tweaks

amclark42 committed:
Per TAPAS strategy meeting, update color of reader description.

I also added a border to the description, though I think it might
need tweaking at different screen sizes.

amclark42 committed:
Adjust z-index of sticky header so it doesn't hide dropdown menus.

The popups for TAPAS Generic and Hieractivity still appear under
the header.

Also, added LESS variables for medium and large screen widths, and
adjusted the display of the Reader description at smaller screen
sizes.

amclark42 committed:
Add base for "My TAPAS" wireframe.

In contrast with the existing "My TAPAS" page, this version
attempts to communicate the relationship between project,
collection, and core file.

I originally planned to use Bootstrap's "Card" feature, but I
decided to roll my own instead.

amclark42 committed:
Wrap `.tapas-container` heading element in a <div>.

...and move the navigation links up next to that heading, using
Bootstrap's `.nav-tab` class for styling.

I did modify Bootstrap's styles so that the tab outline is always
visible, even if the link isn't active.

I also added border-bottom to links, rather than underlines, so
that we're not violating WCAG principles by relying only on color
to tell people that these are links.

amclark42 committed:
Add a few Bootstrap icons

amclark42 committed:
Add general Javascript; tweak styling.

The tapas.js script currently only monitors
`.tapas-container-content` (if it exists) and adds a class to its
parent when Bootstrap collapses it. This lets us style the heading
row appropriately.

amclark42 committed:
As suggested by @candyhazlett, use tables to represent TEI records

amclark42 committed:
Convert LESS file to SCSS

amclark42 committed:
Include information on components that are repeated across files

amclark42 committed:
Use "application" layout; set up outer structure of TAPAS webpage.

Added ERB files for the TAPAS header and footer. The XHTML is
rendering, though now I need to get familiar with the web assets
system. (Which is to say: the site is not pretty.)

amclark42 committed:
Use the ".html.erb" file extension for new templates

amclark42 committed:
With @candyhazlett, update web assets manifest

amclark42 committed:
Commit clobbered public/assets (done with @candyhazlett)

amclark42 committed:
Merge branch 'develop' of ../tapas-wireframes_CUT into web-display.

Filtered tapas-wireframes repository to only the new web assets.

amclark42 committed:
Move web assets to where they should live in Rails.

Deleted the wireframe-specific assets, which I forgot to filter out
(oops).

amclark42 committed:
Add general TAPAS JS to application.js.

Temporarily remove CSS from output; I haven't gotten to it yet.

amclark42 committed:
Edit imports in application.css.scss; fold in new styles

amclark42 committed:
Delete some unused CSS and SCSS files.

amclark42 committed:
Use logo from Rails assets

amclark42 committed:
Create new partial for the user navbar in the header.

amclark42 committed:
Clean up routes a little.

Removed a couple routes I know we're getting rid of. Condensed the
routes for Projects, Collections, and Core Files. Also, added `:`
to identifiers in the route path definitions.

Eventually we'll need more complex URLs for projects, etc., but it
looks like Rails 7 has more powerful routing capabilities. It might
be good to wait until we've upgraded.

amclark42 committed:
With @candyhazlett, remove upsert routes.

The upsert routes were Drupal-specific, and probably won't be
needed going forward. This means we can use the default
`resources :projects` routes provided by Rails.

amclark42 committed:
Clean up commented-out code in routes.rb

amclark42 committed:
Merge branch 'develop' into web-display

Fold in @candyhazlett's work on PR #93

amclark42 committed:
Add Javascript handlers for expanding or collapsing all containers.

Changed the primary button color to something more befitting TAPAS.
(Updates from tapas-wireframes, commits 0f833d0 through 501a0a2.)

amclark42 committed:
Refine CSS styles for Browse and My TAPAS pages.

The Browse page uses a terse list format, while My TAPAS is similar
but can make use of more vertical space. With these changes, I can
start replacing the Blacklight index templates.

Changes from tapas-wireframes repository, through commit 6316f64.

amclark42 committed:
Update styling and behavior of TAPAS containers, links.

Reflects changes in tapas-wireframes repository through commit
`af865ec`.

amclark42 committed:
Make sure nested lists in deploy guide display properly.

GitHub requires us to use four spaces for nested items, not 2.

Also, updated the branch we recommend using.

amclark42 committed:
Update filename of TAPAS logo used as a placeholder thumbnail.

amclark42 committed:
First attempt at "Browse Projects" page.

Currently, only the first project is shown. Rendering all 25
projects takes a little less than a minute on my computer. I'm
certain there are improvements I can make.

amclark42 committed:
Add short browse method to the projects controller.

The browse method absolutely needs more work.

Rather than using a partial on the projects collection, I took
inspiration from the Archive of Our Own codebase and iterated
through each project, rendering the partial for each. It seems to
work best, but still takes a minute to load.

amclark42 committed:
Set up public method to determine if a project is publicly visible.

I wanted to use Rails `scope`s, but they yield an "undefined
method" error when I try.

The placeholder method and eager loading provide a slight
improvement in page load times.

amclark42 committed:
Fix image filename; use `tag.a` instead of `content_tag`

amclark42 committed:
Add date updated

amclark42 committed:
Committing code that doesn't work for experimentation purposes.

May work on @candyhazlett's machine?? We're experimenting

amclark42 committed:
Move collapse toggle button + SVG to its own shared partial

amclark42 committed:
Define the HTML ID for the project's collapsible.

The ID is the string "collapsible-" plus the value of
`item_counter`. With these changes, the Browse Projects page now
starts as a plain list of all projects, with additional information
folded away.

amclark42 committed:
Allow the collapse button to include a textual description

amclark42 committed:
Rename directory for control components.

I started `_collapse_button.html.erb` as just the SVG, but it's
more useful to consider that partial a "control", encompassing the
complex attributes that need to be placed to describe the component
for screen readers and other assistive technology.

amclark42 committed:
Merge branch 'develop' into web-display

amclark42 committed:
Cleanup

candyhazlett commented on LOCAL_DEPLOY_GUIDE.md:
```diff
@@ -3,47 +3,50 @@ Local Deploy Guide
+    2. Go to the core configuration directory: `cd server/solr/tapas-core/conf`
+    3. Create a configuration file: `touch solrconfig.xml`
+    4. Create a schema file: `touch schema.xml`
+    5. Populate each file with the contents of the files received with the same names using the text editor of your choice. **??? Can we make this easier?**
```
This could be simplified by including the full path for the config files as arguments in the `touch` command.

candyhazlett replied:
Then the user remains in the topmost directory and won't have to navigate to a different directory to restart Solr, which they should do after updating the configuration and schema.

amclark42 replied:
Good idea. I can make that change.

My question was specifically for item 5. I didn't receive any files to use in that step, and I ended up using old copies from TAPAS Prod.

Separately from this PR, I'd like us to consider committing `solrconfig.xml` and `schema.xml` files to the repository, which would make setup easier and reduce the overhead of keeping our local files in sync. I don't *think* those files pose security risks, though I could be wrong. If they *are* security risks and shouldn't be committed, step 5 should probably include information on where to find the canonical version of those files. (Not via link, but certainly where to look.)

candyhazlett replied:
@amclark42,  aside from the name of the core itself, the config files don't contain any custom or potentially insecure data. So it's _probably_ fine for them to exist somewhere public in the TAPAS Github universe. Though I would prefer using a Gist for each file, or something like that. There's also the iminent Dockerization, which will occur pre-launch. Once that happens, the user won't have to manually configure Solr at all.

amclark42 replied:
@candyhazlett Can you say more about what a Gist would buy us here?

We'd have the same problem when setting up the Docker image. If we commit the Dockerfile here (like I did for the [Word Vector Interface](https://github.com/NEU-DSG/word-vector-interface/blob/main/docker/Dockerfile)), we'll just end up sharing the link to the Gist. It'll be as public as if we committed the files here. If we _don't_ commit the Dockerfile, or implement a series of secrets to get at the configurations, we're adding manual steps that make the running and installation hard to do.

I think it makes the most sense to tie the configuration files to the code that relies on that setup, rather than adding a step. It would allow us to tune the configuration without requiring us to track separately which version works on which branch. And it would make it easier to do Dockerization later too.

candyhazlett commented on app/controllers/projects_controller.rb:
```diff
@@ -8,7 +8,12 @@ class ProjectsController < ApplicationController
-
+  
+  def browse
+    @projects = Project.publicly_visible.includes(:collections, :core_files)
```
This is great! `.includes` is super efficient.

amclark42 replied:
Thanks!

candyhazlett reviewed (changes requested):
## 📝 Basic Info
| Items                       | Details                                                                                                                                     |
|-----------------------------|---------------------------------------------------------------------------------------------------------------------------------------------|
| **PR Title**                | Replace default application layout and homepage (currently Browse Projects) [103](https://github.com/NEU-DSG/tapas_rails/pull/103)                                                                |
| **Related Issue/Ticket**    |  N/A, no issue created                                                                                                                  |
| **Technical Specification** | [Browse TAPAS content](https://docs.google.com/document/d/1vsD_ZJtoHAQo5-T-frH3w64Tpbmuyp4y_JaVu01Hc6w/edit?tab=t.0#heading=h.v90ktn7h1hk4) |

## 🎯 Initial steps
- [x] **Read PR Description** - linked above in **Basic Info**
- [x] **Read Requirements** 
  - ~~[ ] Github issue #~~
  - [x] Functional Specifications doc section,"Browse TAPAS content", linked above in **Basic Info**
    - Per the description, this PR is based on wireframes created by the PR author. No Github issue exists, so for the purpose of this review, acceptance criteria are taken from the relevant section of the Functional Specifications.
- [x] **Branch Checkout to verify:**
  - [x] PR builds successfully
  - ~~Automated tests pass (if any)~~ N/A

## 📋 Requirements Verification
### Acceptance Criteria (AC) Status
#### AC1: "Readers coming to the site can use a “Browse TAPAS” or “Explore TAPAS” space to get an overview of the TAPAS materials"
- [x] **Status:** ⚠️ Partially Met
- [x] **Verification:** Local
- [x] **Testing:** Manual
  - Results:
    - started the Rails server, then visited http://localhost:3000 and clicked the 'Browse' nav menu link
    - clicking the nav menu link loads 'Browse Projects' page
    - link text says 'Browse' rather than “Browse TAPAS” or “Explore TAPAS” as stated in the Specification, or other text that would indicate what the user will be browsing

#### AC2: "See a list or grid view of all TAPAS projects"
- [x] **Status:** ✅ Met
- [x] **Verification:** Local
- [x] **Testing:** Manual
  - Results:
    - list of projects displays on the 'Browse Projects' page

#### AC3: "...all TAPAS projects...Sortable"
- [x] **Status:** ❌ Unmet
- [x] **Verification:** Local
- [x] **Testing:** Manual
  - Test results:
    - 'Sort' control appears and displays options for sorting, as expected
    - selecting the 'alphabetically' sort option, then clicking 'OK', does not sort the projects list alphabetically

#### AC4: "...all TAPAS projects... filterable by owner, by title, by date created, by number of TEI files"
- [x] **Status:** ❌ Unmet
- [x] **Verification:** Local
- [x] **Testing:** Manual
  - Test results:
    - 'Filter' control does not appear on the 'Browse Projects' page

## 🔍 Code Quality Assessment
### 📁 Files Changed:
- [ ] **Red flags?**
- [ ] **Obvious bugs?**
- [ ] **Obvious regressions?**
- [x] **Missing error handling?**

### Structure & Design
- **Concerns/Issues:**
  - [x] Single function doing multiple tasks
    - `prepareTapas()` handles 3 different event listener setups
  - **Where?:** `tapas.js:63`
  - **Suggestion:** Extract tasks out to distinct functions.

  - [x] Code duplication
    - `collapseAllDisclosures` and `expandAllDisclosures` are nearly identical
  - **Where?:** `tapas.js:26`
  - **Suggestion:** Refactor to remove duplication.

### ⚙️ Technical Implementation
- **Concerns/Issues:**
  - [x] Potential for runtime error if Bootstrap isn't available `bootstrap.Collapse.getOrCreateInstance(myTarget).hide();`
  - **Where?:** `tapas.js:30, 42`
  - **Suggestion:**  
    - Verify `bootstrap` global exists in advance
    - and/or add error handling for it and other external dependencies

  - [x] Using an instance method `Project#publicly_visible` for class method behavior is convenient, but also a Rails anti-pattern. i.e., it goes against Rails conventions, especially since it's a straightforward where clause so the scope is appropriate and should work 
  - **Where?:** `project.rb:209`
  - **Suggestion:**  
    - We figure out why the scope you defined initially `scope :publicly_visible, -> { where(is_public: true) }` kept returning as `undefined`, then add it back instead of working around it, creating more tech debt
    - Add indexes to Project on `is_public` to speed up the query, which will be important when we have the full db back online with way more than the 25 projects generated from the dummy data rake task

### ~~🔒 Security Considerations~~N/A

### 📚 Documentation
- [x] Code comments added where necessary
- [x] Logic is described well
~~- [ ] API documentation updated (for TAPAS-Xq)~~

### ⚡ Performance & Scalability
- [x] No impact on performance

## 📝 Feedback
**Questions from PR Description:**
- **Q**: Does the homepage load as expected?
  - **A**: Yes, the current homepage (defined as the Browse Projects page in the PR description) loads without issue
- **Q**: Are there any places where the TAPAS layout isn't working, or the page content is negatively affected?
  - **A**: See AC3.

**Specific Feedback Requested:**
- **Q**: Have I made any errors?
  - **A**: a copy/paste error in ```tapas.js```:
      - ```expandAllDisclosures()``` has wrong end comment: ```// end collapseAllDisclosures()```
- **Q**: Are there things I can improve?
  - **A**: Yes
    1. (on a different PR) let's convert the view templates created in this PR from ```.erb``` to ```.haml``` format
       - there's an online conversion [tool](https://awsm-tools.com/html-to-haml)
       - and a ruby [gem](https://rubygems.org/gems/html2haml) library. I'd say the conversion tool is the way to go. It will take longer but it's safer than the bulk action of the gem library
    2. In  ```reader.js```, there's a few DOM element queries ```document.querySelector``` sprinkled throughout that we could cache using a ```const``` at initialization. That said, we could also use some error handling for if/when any elements are missing from the DOM.

### ✨ What's Working Well
- The user interface looks amazing
- Good use of semantic naming for view component IDs and classes
- Good namespacing for JS functions
- Excellent documentation/comments
- Collapse/Expand controls work perfectly 

### 🐛 Decision
- [ ] ✅ **APPROVED** - Ready to merge
- [x] 🔄 **CHANGES REQUESTED** - See summary below
- [ ] ❌ **NEEDS MAJOR REWORK** - Significant issues found

**Summary:** 
- Functionality implemented by the PR is solid overall, minus the four issues outlined in the Structure & Design, and Technical Implementation sections.
- AC1 and AC3 should be addressed before merging, either with additional code/components or in the PR description if you opt not to address them in this PR.
- Whatever the plan is for implementing AC4, since this PR moves the codebase away from the Blacklight Catalog layout's built-in filter components and logic the app relied on previously, it should be documented along with what led to the decision as it constitutes a major departure.

**Additional Notes**
- Questions for the author
  - Would you mind documenting the modifications to the asset pipeline, as represented by the changes made in app/assets/javascripts/application.js?
  - Could you also add a link to the Browse page wireframe? I can't find it in the Dropbox.

## ✍️ Sign-off
| Field                          | Value                                           |
|--------------------------------|-------------------------------------------------|
| **Additional Review Required** | yes, once requested changes are done                              |
| **Time Spent:**                | > 1 hr                                          |
| **Most Time Consuming:**       | code review due to extensive commit history     |

candyhazlett reviewed (changes requested):
## 📝 Basic Info
| Items                       | Details                                                                                                                                     |
|-----------------------------|---------------------------------------------------------------------------------------------------------------------------------------------|
| **PR Title**                | Replace default application layout and homepage (currently Browse Projects) [103](https://github.com/NEU-DSG/tapas_rails/pull/103)          |
| **Related Issue/Ticket**    | N/A                                                                                                                                         |
| **Technical Specification** | [Browse TAPAS content](https://docs.google.com/document/d/1vsD_ZJtoHAQo5-T-frH3w64Tpbmuyp4y_JaVu01Hc6w/edit?tab=t.0#heading=h.v90ktn7h1hk4) |

## 🎯 Initial steps
- [x] **Read PR Description** - linked above in **Basic Info**
- [x] **Read Requirements** 
  - ~~[ ] Github issue #~~
  - [x] Technical Specifications doc section,"Browse TAPAS content", linked above in **Basic Info**
    - Per the description, this PR is based on wireframes created by the PR author. No Github issue exists, so for the purpose of this review, acceptance criteria are taken from the relevant section of the Technical Specifications.
- [x] **Branch Checkout to verify:**
  - [x] PR builds successfully
  - ~~Automated tests pass (if any)~~ N/A

## 📋 Requirements Verification
### Acceptance Criteria (AC) Status
#### AC1: "Readers coming to the site can use a “Browse TAPAS” or “Explore TAPAS” space to get an overview of the TAPAS materials"
- [x] **Status:** ⚠️ Partially Met
- [x] **Verification:** Local
- [x] **Testing:** Manual
  - Results:
    - started the Rails server, then visited http://localhost:3000 and clicked the 'Browse' nav menu link
    - clicking the nav menu link loads 'Browse Projects' page
    - link text says 'Browse' rather than “Browse TAPAS” or “Explore TAPAS” as stated in the Specification, or other text that would indicate what the user will be browsing

#### AC2: "See a list or grid view of all TAPAS projects"
- [x] **Status:** ✅ Met
- [x] **Verification:** Local
- [x] **Testing:** Manual
  - Results:
    - list of projects displays on the 'Browse Projects' page

#### AC3: "...all TAPAS projects...Sortable"
- [x] **Status:** ❌ Unmet
- [x] **Verification:** Local
- [x] **Testing:** Manual
  - Test results:
    - 'Sort' control appears and displays options for sorting, as expected
    - selecting the 'alphabetically' sort option, then clicking 'OK', does not sort the projects list alphabetically

#### AC4: "...all TAPAS projects... filterable by owner, by title, by date created, by number of TEI files"
- [x] **Status:** ❌ Unmet
- [x] **Verification:** Local
- [x] **Testing:** Manual
  - Test results:
    - 'Filter' control does not appear on the 'Browse Projects' page

## 🔍 Code Quality Assessment
### 📁 Files Changed:
- [ ] **Red flags**
- [ ] **Obvious bugs**
- [ ] **Obvious regressions**
- [x] **Missing error handling**

### Structure & Design
- **Concerns/Issues:**
  - [x] Single function doing multiple tasks
    - `prepareTapas()` handles 3 different event listener setups
  - **Where?:** `tapas.js:63`
  - **Suggestion:** Extract tasks out to distinct functions.

  - [x] Code duplication
    - `collapseAllDisclosures` and `expandAllDisclosures` are nearly identical
  - **Where?:** `tapas.js:26`
  - **Suggestion:** Refactor to remove duplication.

### ⚙️ Technical Implementation
- **Concerns/Issues:**
  - [x] Potential for runtime error if Bootstrap isn't available `bootstrap.Collapse.getOrCreateInstance(myTarget).hide();`
  - **Where?:** `tapas.js:30, 42`
  - **Suggestion:**  
    - Verify `bootstrap` global exists in advance
    - and/or add error handling for it and other external dependencies

  - [x] Using an instance method `Project#publicly_visible` for class method behavior is convenient, but also a Rails anti-pattern. i.e., it goes against Rails conventions, especially since it's a straightforward where clause so the scope is appropriate and should work 
  - **Where?:** `project.rb:209`
  - **Suggestion:**  
    - We figure out why the scope you defined initially `scope :publicly_visible, -> { where(is_public: true) }` kept returning as `undefined`, then add it back instead of working around it, creating more tech debt
    - Add indexes to Project on `is_public` to speed up the query, which will be important when we have the full db back online with way more than the 25 projects generated from the dummy data rake task

### ~~🔒 Security Considerations~~N/A

### 📚 Documentation
- [x] Code comments added where necessary
- [x] Logic is described well
~~- [ ] API documentation updated (for TAPAS-Xq)~~

### ⚡ Performance & Scalability
- [x] No impact on performance

## 📝 Feedback
### ✨ What's Working Well
- The user interface looks amazing
- Good use of semantic naming for view component IDs and classes
- Good namespacing for JS functions
- Excellent documentation/comments
- Collapse/Expand controls work perfectly

**Questions from PR Description:**
- Q: Does the homepage load as expected?
  - A: Yes, the current homepage (defined as the Browse Projects page in the PR description) loads without issue
- Q: Are there any places where the TAPAS layout isn't working, or the page content is negatively affected?
  - A: See AC3 in the Acceptance Criteria (AC) Status section above

**Specific Feedback Requested:**
- Q: Have I made any errors?
  - A: AC1 and AC3 are omissions, rather than errors.
    - a copy/paste error in ```tapas.js```:
      - ```expandAllDisclosures()``` has wrong end comment: ```// end collapseAllDisclosures()```
- Q: Are there things I can improve?
  - A: Yes
    1. (on a different PR) let's convert the view templates created in this PR from `.erb` to `.haml` format for ease of use.
       - there's an online conversion [tool](https://awsm-tools.com/html-to-haml)
       - and a ruby [gem](https://rubygems.org/gems/html2haml) library. I'd say the conversion tool is the way to go. It will take longer but it's safer than the bulk action of the gem library
    2. In  ```reader.js```, there's a few DOM element queries ```document.querySelector``` sprinkled throughout that we could cache using a ```const``` at initialization. That said, we could also use some error handling should any of the elements be missing from the DOM.

### 🐛 Decision
- [ ] ✅ **APPROVED** - Ready to merge
- [x] 🔄 **CHANGES REQUESTED** - See summary below
- [ ] ❌ **NEEDS MAJOR REWORK** - Significant issues found

**Summary:** 
- Functionality implemented by the PR is solid overall, minus AC3 and the four issues outlined in the Structure & Design, and Technical Implementation sections.
- AC1 and AC3 should be addressed before merging, either with additional code/components or in the PR description if you opt not to address them in this PR.
- Whatever the plan is for implementing AC4, since this PR moves the codebase away from the Blacklight Catalog layout's built-in filter components and logic the app relied on previously, it should be documented along with what led to the decision as it constitutes a major departure.

**Additional Notes**
- Questions for the author
  - Would you mind documenting the modifications to the asset pipeline, as represented by the changes made in app/assets/javascripts/application.js?
  - Could you also add a link to the Browse page wireframe? I can't find it in the Dropbox.

## ✍️ Sign-off
| Field                          | Value                                       |
|--------------------------------|---------------------------------------------|
| **Additional Review Required** | yes, once requested changes have been made  |
| **Recommended Reviewers**      | @candyhazlett                               |
| **Time Spent:**                | > 1 hr                                      |
| **Most Time Consuming:**       | code review due to extensive commit history |

amclark42 commented:
I feel the changes requested are out of the scope of this PR's development work, but I agree that the PR description can be more clear about its scope, and the current state of the Browse Projects page. I will update the description, and re-request a review at that time.

In particular, @candyhazlett's acceptance criterion 3 (sorting) will be implemented in a future development push. The TAPAS strategy group decided not to pursue criterion 4 (filtering) in our [April 9, 2025 meeting](https://docs.google.com/document/d/1H64qh1v5QYHuPFuQ24k-NnBJJAxCVzopLWYOdFueF0M/edit?tab=t.0#heading=h.3z4vnfz1kkqo). The functional specification will be updated.

Besides updating the description, I will fix these issues:
- "`collapseAllDisclosures` and `expandAllDisclosures` are nearly identical"
    - "`expandAllDisclosures()` has wrong end comment: `// end collapseAllDisclosures()`"
- "Potential for runtime error if Bootstrap isn't available"
- "we could also use some error handling should any of the elements be missing from the DOM"

amclark42 commented:
@candyhazlett — Some comments on your other feedback:

> link text says 'Browse' rather than “Browse TAPAS” or “Explore TAPAS” as stated in the Specification, or other text that would indicate what the user will be browsing

I'm inclined to leave the navigation link as-is for the moment.

It's an artifact of the way that the Browse page worked in Drupal (the Rails Blacklight catalog too) — there was only one HTML page, with Javascript navigating you between the projects, collections, and TEI records. So you always landed on "Projects" first, and you'd click a button from there to see one of the other views. "Browse TAPAS" was a useful enough way to describe the three-headed nature of the URL address.

By changing to three separate URLs for each "Browse" type, we make it easy for people to navigate directly to the list they want. But we haven't yet thought about how to handle this in the navbar. My thought is a submenu (Browse > Projects), but I want to run it past the strategy team, possibly with a wireframe example.

> `prepareTapas()` handles 3 different event listener setups

I don't think that's a problem — all of those event listeners are setting up the expand and collapse controls. They belong together, and the JS doesn't do anything else at the moment. For future-proofing purposes, however, I'll rename the function. That will leave open the possibility that this function can be called by a future `this.prepareTapas()` (which does other things besides setting up collapsibles).

> figure out why the scope you defined initially `scope :publicly_visible, -> { where(is_public: true) }` kept returning as `undefined`, then add it back instead of working around it

Yes, that is something we should do. I would rather work on that after this PR is accepted.

> (on a different PR) let's convert the view templates created in this PR from `.erb` to `.haml` format for ease of use.

Let's discuss if/when we should do a conversion in a working meeting. ERB and HAML both output to XHTML, so that's not a factor. I do recognize that you probably prefer HAML and find it easier to read. However, I'm an XML person — I *like* tags to close, I find that grounding. ERB works better for me because I can, at a glance, understand what is Ruby and what corresponds to the elements I see in my browser inspector. In other words, HAML's "ease of use" would be the exact opposite for me. If I'm going to be the one iterating over these pages, in response to user testing and future development, I'd rather keep the files in ERB format for quicker turnaround.

> In `reader.js`, there's a few DOM element queries `document.querySelector` sprinkled throughout that we could cache using a `const` at initialization.

I prefer not to stash DOM elements in variables at instantiation unless absolutely necessary. It's too easy for the DOM to change under you; I want to be positive that I'm getting the elements as they are when the function is called.

I'm also leery of using `const` unless I'm absolutely sure its value isn't going to change — I don't use it for existing DOM nodes. I'd rather use `let` and `var`, carefully scoped at the function level.

> Whatever the plan is for implementing AC4, since this PR moves the codebase away from the Blacklight Catalog layout's built-in filter components and logic the app relied on previously, it should be documented along with what led to the decision as it constitutes a major departure.
> Would you mind documenting the modifications to the asset pipeline, as represented by the changes made in app/assets/javascripts/application.js?

I will expand on the removal in the PR description. The description already states that the Blacklight Catalog view had stopped working, but it doesn't say that that was specifically because it doesn't work with XHTML output. I will add that.

You're right that this is a major departure, but it's also an expected one. The Blacklight Catalog was always very much a placeholder, IMO. We were always going to redesign the Browse pages to better represent the TAPAS resources.

> Could you also add a link to the Browse page wireframe? I can't find it in the Dropbox.

Can do. For future reference, the wireframes are in [this (private) repository](https://github.com/NEU-DSG/tapas-wireframes), and [a copy is up on TAPAS Prod](https://tapasproject.org/wireframes/wireframe-index.html).

candyhazlett commented:
> In particular, @candyhazlett's acceptance criterion 3 (sorting) will be implemented in a future development push. 

@amclark42, would you mind creating a Github issue for the sort feature?

amclark42 commented:
> > In particular, @candyhazlett's acceptance criterion 3 (sorting) will be implemented in a future development push.
> 
> @amclark42, would you mind creating a Github issue for the sort feature?

Will do.

amclark42 committed:
Test for Bootstrap; add generic function to toggle collapsibles.

amclark42 committed:
Clean up tapas.js.

Briefly tried removing `collapseAllDisclosures()` and
`expandAllDisclosures()` in favor of an anonymous function calling
`toggleAllDisclosures()`. I think the convenience functions are
more readable when called, so they remain.

amclark42 commented:
@candyhazlett — The PR description now has a section "2025-08 edit: Additional context" which contains more information on the scope of this PR. I also patched in two commits from the `tapas-wireframes` repository which address some of your suggestions for `tapas.js`.

candyhazlett commented on app/assets/javascripts/tapas.js:
```diff
@@ -25,10 +25,7 @@ tapas.general = {};
     console.log("Collapsing containers");
-    document.querySelectorAll('.tapas-container button[data-bs-toggle="collapse"][aria-expanded="true"]').forEach(element => {
-      let myTarget = document.querySelector(element.dataset.bsTarget);
-      bootstrap.Collapse.getOrCreateInstance(myTarget).hide();
```
Great!
